File: sort/c/heap_sort.c
/* 
 * heap_sort.c
 * Perform heap sort on an array of integers
 */

#include "sort.h"

static void swap(int *a, int *b) {
    int temp;
    temp = *a;
    *a = *b; 
    *b = temp;
}

static void sift_down(int arr[], int parent_index, int last_index) {
    int left = 2 * parent_index;
    int right = 2 * parent_index + 1;
    int largest;

    if(left <= last_index && arr[left] > arr[parent_index])
        largest = left;
    else
        largest = parent_index;
    
    if(right <= last_index && arr[right] > arr[largest]) 
        largest = right;

    if(largest != parent_index) {
        swap(&arr[parent_index], &arr[largest]);
        sift_down(arr, largest, last_index);
    }
}

static void build_heap(int arr[], int last_index) {
    int parent_index; 
    for(parent_index = (last_index/2)-1; parent_index >= 0; parent_index--) {
        sift_down(arr, parent_index, last_index);
    }
}

void heap_sort(int arr[], int len) {
    int i, last_index; 
    last_index = len - 1;
    /* first create a max-heap from the array */
    build_heap(arr, last_index);
    /*now take elements from heap*/
    for(i = last_index; i > 0; i--) {
        swap(&arr[0], &arr[i]); 
        last_index -= 1; 
        sift_down(arr, 0, last_index);
    }
}

File: sort/java/HeapSort.java
/** 
 * This class implements HeapSort on an array of integers. 
 * The advantage of heapsort is it is an in-place algorithm,
 * meaning it overwrites its input with output
 * Time complexity for heap sort is O(nlogn)
 */

public class HeapSort {
    private static void swapArrayElements(int [] arr, int index1, int index2) {
        if(index1 > arr.length || index1 < 0 || index2 > arr.length || index2 < 0) {
            throw new java.lang.IndexOutOfBoundsException(); 	
        }
        int temp = arr[index1];
        arr[index1] = arr[index2];
        arr[index2] = temp;
    }

    private static void siftDown(int [] arr, int parentIndex, int lastIndex) {
        int left = 2*parentIndex;
        int right = 2*parentIndex+1;
        int largest;

        if(left <= lastIndex && arr[left] > arr[parentIndex]) {
            largest = left;
        } else {
            largest = parentIndex;
        }

        if(right <= lastIndex && arr[right] > arr[largest]) {
            largest = right;
        }

        if(largest != parentIndex) {
            swapArrayElements(arr, parentIndex, largest);
            siftDown(arr, largest, lastIndex);
        }
    }

    private static void buildHeap(int [] arr, int lastIndex) 
    {
        for(int parentIndex = (lastIndex/2)-1; parentIndex >= 0; parentIndex--) {
            siftDown(arr, parentIndex, lastIndex);
        }
    }
 
    public static void sort(int [] arr) 
    {
        int lastIndex = arr.length - 1;
        //first create a max-heap from the array
        buildHeap(arr, lastIndex);

        for(int i = lastIndex; i > 0; i--) {
            swapArrayElements(arr, 0, i);
            lastIndex -= 1;
            siftDown(arr, 0, lastIndex);
        }
    }
}
